java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java

```diff
@@ -0,0 +1,991 @@
+  public int getValueCapacity() {
+    return getValidityAndOffsetValueCapacity();
+  }
+  
+  protected int getOffsetBufferValueCapacity() {
+    return checkedCastToInt(offsetBuffer.capacity() / OFFSET_WIDTH);
+  }
+  
+  private int getValidityAndOffsetValueCapacity() {
+    final int offsetValueCapacity = Math.max(getOffsetBufferValueCapacity() - 1, 0);
+    return Math.min(offsetValueCapacity, getValidityBufferValueCapacity());
+  }
+
+  private int getValidityBufferValueCapacity() {
+    return capAtMaxInt(validityBuffer.capacity() * 8);
+  }
+
+  /**
+   * Sets the list at index to be not-null.  Reallocates validity buffer if index
+   * is larger than current capacity.
+   */
+  public void setNotNull(int index) {
+    while (index >= getValidityAndOffsetValueCapacity()) {
+      reallocValidityAndOffsetBuffers();
+    }
+    BitVectorHelper.setBit(validityBuffer, index);
+    lastSet = index;
+  }
+
+  /**
+   * Start a new value in the list vector.
+   *
+   * @param index index of the value to start
+   */
+  public long startNewValue(long index) {
+    while (index >= getValidityAndOffsetValueCapacity()) {
+      reallocValidityAndOffsetBuffers();
+    }
+    for (long i = lastSet + 1; i <= index; i++) {
+      final long currentOffset = offsetBuffer.getLong(i * OFFSET_WIDTH);
+      offsetBuffer.setLong((i + 1) * OFFSET_WIDTH, currentOffset);
+    }
+    BitVectorHelper.setBit(validityBuffer, index);
+    lastSet = index;
+    return checkedCastToInt(offsetBuffer.getLong((lastSet + 1) * OFFSET_WIDTH));
+  }
+
+  /**
+   * End the current value.
+   *
+   * @param index index of the value to end
+   * @param size  number of elements in the list that was written
+   */
+  public void endValue(int index, int size) {
+    final long currentOffset = offsetBuffer.getLong((index + 1) * OFFSET_WIDTH);
+    offsetBuffer.setLong((index + 1) * OFFSET_WIDTH, currentOffset + size);
+  }
+
+  /**
+   * Sets the value count for the vector.
+   *
+   * @param valueCount   value count
+   */
+  @Override
+  public void setValueCount(int valueCount) {
+    this.valueCount = valueCount;
+    if (valueCount > 0) {
+      while (valueCount > getValidityAndOffsetValueCapacity()) {
+        /* check if validity and offset buffers need to be re-allocated */
+        reallocValidityAndOffsetBuffers();
+      }
+      for (long i = lastSet + 1; i < valueCount; i++) {
+        /* fill the holes with offsets */
+        final long currentOffset = offsetBuffer.getLong(i * OFFSET_WIDTH);
+        offsetBuffer.setLong((i + 1) * OFFSET_WIDTH, currentOffset);
+      }
+    }
+    /* valueCount for the data vector is the current end offset */
+    final long childValueCount = (valueCount == 0) ? 0 :
+            offsetBuffer.getLong((lastSet + 1) * OFFSET_WIDTH);
+    /* set the value count of data vector and this will take care of
+     * checking whether data buffer needs to be reallocated.
+     */
+    vector.setValueCount(checkedCastToInt(childValueCount));
```

Comment:
Also indicate this in the javadoc here and probably for the class. It might even be a good idea to raise an error if the user tries to add too many elements, otherwise things might just start looking wrong.

Reply:
done